langchain/src/chat_models/googlepalm.ts

```diff
@@ -0,0 +1,242 @@
+import { DiscussServiceClient, protos } from "@google-ai/generativelanguage";
+import { GoogleAuth } from "google-auth-library";
+import { CallbackManagerForLLMRun } from "../callbacks/manager.js";
+import { AIMessage, BaseMessage, ChatResult } from "../schema/index.js";
+import { getEnvironmentVariable } from "../util/env.js";
+import { BaseChatModel, BaseChatModelParams } from "./base.js";
+
+export interface GooglePalmChatInput extends BaseChatModelParams {
+  /**
+   * Model Name to use
+   *
+   * Note: The format must follow the pattern - `models/{model}`
+   */
+  model?: string;
+
+  /**
+   * Controls the randomness of the output.
+   *
+   * Values can range from [0.0,1.0], inclusive. A value closer to 1.0
+   * will produce responses that are more varied and creative, while
+   * a value closer to 0.0 will typically result in less surprising
+   * responses from the model.
+   *
+   * Note: The default value varies by model
+   */
+  temperature?: number;
+
+  /**
+   * Top-p changes how the model selects tokens for output.
+   *
+   * Tokens are selected from most probable to least until the sum
+   * of their probabilities equals the top-p value.
+   *
+   * For example, if tokens A, B, and C have a probability of
+   * .3, .2, and .1 and the top-p value is .5, then the model will
+   * select either A or B as the next token (using temperature).
+   *
+   * Note: The default value varies by model
+   */
+  topP?: number;
+
+  /**
+   * Top-k changes how the model selects tokens for output.
+   *
+   * A top-k of 1 means the selected token is the most probable among
+   * all tokens in the model’s vocabulary (also called greedy decoding),
+   * while a top-k of 3 means that the next token is selected from
+   * among the 3 most probable tokens (using temperature).
+   *
+   * Note: The default value varies by model
+   */
+  topK?: number;
+
+  examples?: protos.google.ai.generativelanguage.v1beta2.IExample[];
+
+  /**
+   * Google Palm API key to use
+   */
+  apiKey?: string;
+}
+
+export class ChatGooglePalm
+  extends BaseChatModel
+  implements GooglePalmChatInput
+{
+  get lc_secrets(): { [key: string]: string } | undefined {
+    return {
+      apiKey: "GOOGLEPALM_API_KEY",
+    };
+  }
+
+  model = "models/chat-bison-001";
+
+  temperature?: number = undefined; // default value chosen based on model
+
+  topP?: number = undefined; // default value chosen based on model
+
+  topK?: number = undefined; // default value chosen based on model
+
+  examples: protos.google.ai.generativelanguage.v1beta2.IExample[] = [];
+
+  apiKey?: string;
+
+  private client: DiscussServiceClient;
+
+  constructor(fields?: GooglePalmChatInput) {
+    super(fields ?? {});
+
+    this.model = fields?.model ?? this.model;
+    if (this.model && !this.model.startsWith("models/")) {
```

Comment:
If the prefix is always there, do we really need it?

Reply:
Since the PaLM API requires model name as `models/{model}`, I guess, keeping it as close as possible with underlying implementation makes better sense, since if Google changes the naming pattern later. I'll remove this check for `modelName.startsWith('models/')`